packages/contracts/src/VaultPlugin.sol

```diff
@@ -0,0 +1,157 @@
+
+    /// @notice A boolean flag to indicate if transfers are enabled or not
+    bool public transfersEnabled;
+
+    /// @dev Emitted when transfers are disabled
+    error TransfersDisabled();
+
+    /// @notice Initializes the plugin when build 1 is installed.
+    /// @param _asset Set the underlying asset contract.
+    function initialize(IDAO _dao, IERC20MetadataUpgradeable _asset) external initializer {
+        __PluginUUPSUpgradeable_init(_dao);
+        __ERC4626_init(_asset);
+    }
+
+    /// @notice Toggles the state of transfers
+    /// @dev This function can only be called by an account with the `TOGGLE_TRANSFER_PERMISSION_ID` permission
+    function toggleTransfers() external auth(TOGGLE_TRANSFER_PERMISSION_ID) {
+        transfersEnabled = !transfersEnabled;
+    }
+
+    /// @notice Returns the total assets held by the DAO
+    /// @dev See {IERC4626-totalAssets}.
+    function totalAssets() public view virtual override returns (uint256) {
+        return IERC20MetadataUpgradeable(asset()).balanceOf(address(dao()));
+    }
+
+    /// @notice Overrides the _deposit function from the ERC4626 contract to deposit assets into the DAO instead of this
+    /// contract.
+    /// @dev See {IERC4626-_deposit}.
+    /// @param caller The address of the caller.
+    /// @param receiver The address of the receiver.
+    /// @param assets The amount of assets to deposit.
+    /// @param shares The amount of shares to issue.
+    function _deposit(address caller, address receiver, uint256 assets, uint256 shares) internal virtual override {
+        // If _asset is ERC777, `transferFrom` can trigger a reenterancy BEFORE the transfer happens through the
+        // `tokensToSend` hook. On the other hand, the `tokenReceived` hook, that is triggered after the transfer,
+        // calls the vault, which is assumed not malicious.
+        //
+        // Conclusion: we need to do the transfer before we mint so that any reentrancy would happen before the
+        // assets are transfered and before the shares are minted, which is a valid state.
+        // slither-disable-next-line reentrancy-no-eth
+        SafeERC20Upgradeable.safeTransferFrom({
+            token: IERC20Upgradeable(asset()),
+            from: caller,
+            to: address(dao()),
+            value: assets
+        });
+        _mint(receiver, shares);
+
+        emit Deposit(caller, receiver, assets, shares);
+    }
+
+    /// @notice Overrides the _withdraw function from the ERC4626 contract to withdraw assets from the DAO instead of
+    /// this contract.
+    /// @dev See {IERC4626-_withdraw}.
+    /// @param caller The address of the caller.
+    /// @param receiver The address of the receiver.
+    /// @param owner The address of the owner.
+    /// @param assets The amount of assets to withdraw.
+    /// @param shares The amount of shares to burn.
+    function _withdraw(
```

Comment:
I think I need a bit of clarity on how we are going to calculate the burn function in association with withdrawing the underlining asset from the DAO.

From my understanding, a donor should be able to withdraw a portion of the DAO asset in relation to the balance of the initial amount they donated to the DAO. Since depositing with this plugin is not the only way a DAO can increase the total amount of asset held by the DAO treasury, I fail to understand how this will achieve this feat

Reply:
The way it works is a donor can withdraw from the DAOs balance in one specific token. This is not necessarily in relation to deposits nor to the DAOs treasury as a whole.

for example If the deposit asset is USDC and the DAO somehow decides to accept USDC for services rendered, them donors vault tokens will increase in value.

another example, the DAO decides to swap 100% of the USDC to USDT, the DAO would effectively be rugging donors.

Reply:
@pythonpete32 so another silly question is, if I as a donor have the ability to withdraw from the DAO treasury is there a way to check for a limit as to how much I can withdraw?

Or do I have the ability to withdraw the whole treasury. 

Reason being that, I don't see that check as to how the regulate what % a donor is permitted to withdraw

Reply:
i defo advise reading the ERC4626 docs on openzeppelin https://docs.openzeppelin.com/contracts/4.x/erc4626, the only logic changed is where the tokens are stored and added pausability on transfers. Also adding a video

https://www.youtube.com/watch?v=L8dijE5qhTg

Reply:
Alright captain. Done. 

Thanks, and PR looks good to me.